qiskit/algorithms/linear_solvers/hhl.py

```diff
@@ -0,0 +1,373 @@
+            observable = [observable]
+
+        expectations = []
+        for circ, obs in zip(observable_circuit, observable):
+            circuit = QuantumCircuit(solution.num_qubits)
+            circuit.append(solution, circuit.qubits)
+            circuit.append(circ, range(nb))
+
+            ob = one_op ^ TensoredOp((nl + na) * [zero_op]) ^ obs
+            expectations.append(~StateFn(ob) @ StateFn(circuit))
+
+        if is_list:
+            # execute all in a list op to send circuits in batches
+            expectations = ListOp(expectations)
+        else:
+            expectations = expectations[0]
+
+        if self._expectation is not None:
+            expectations = self._expectation.convert(expectations)
+
+        if self._sampler is not None:
+            expectations = self._sampler.convert(expectations)
+
+        # evaluate
+        expectation_results = expectations.eval()
+
+        # apply post_processing
+        result = post_processing(expectation_results, nb, scaling)
+
+        return result, expectation_results
+
+    def construct_circuit(self, matrix: Union[np.ndarray, QuantumCircuit],
+                          vector: Union[np.ndarray, QuantumCircuit]) -> QuantumCircuit:
+        """Construct the HHL circuit.
+
+        Args:
+            matrix: The matrix specifying the system, i.e. A in Ax=b.
+            vector: The vector specifying the right hand side of the equation in Ax=b.
+
+        Returns:
+            QuantumCircuit: the QuantumCircuit object for the constructed circuit
+
+        Raises:
+            ValueError: If the input is not in the correct format.
+        """
+        # State preparation circuit - default is qiskit
+        if isinstance(vector, QuantumCircuit):
+            nb = vector.num_qubits
+            vector_circuit = vector
+        elif isinstance(vector, np.ndarray):
+            nb = int(np.log2(len(vector)))
+            vector_circuit = QuantumCircuit(nb)
+            vector_circuit.isometry(vector / np.linalg.norm(vector),
+                                    list(range(nb)), None)
+
+        # If state preparation is probabilistic the number of qubit flags should increase
+        nf = 1
+
+        # Hamiltonian simulation circuit - default is Trotterization
+        if isinstance(matrix, QuantumCircuit):
+            matrix_circuit = matrix
+        elif isinstance(matrix, np.ndarray):
+            if matrix.shape[0] != matrix.shape[1]:
+                raise ValueError("Input matrix must be square!")
+            if np.log2(matrix.shape[0]) % 1 != 0:
+                raise ValueError("Input matrix dimension must be 2^n!")
+            if not np.allclose(matrix, matrix.conj().T):
+                raise ValueError("Input matrix must be hermitian!")
+            if matrix.shape[0] != 2 ** vector_circuit.num_qubits:
+                raise ValueError("Input vector dimension does not match input "
+                                 "matrix dimension!")
```

Comment:
For cases like this and above, it would be nice/helpful to have the failing values in the error message

Reply:
added the values to the error message